### PR TITLE
Injectable mutagen finales, targetable purifier smart shots.

### PR DESCRIPTION
**Pull request by BrianLefler**
Adds a targetable purifier that will remove one targeted mutation, and 1-2 additional untargeted mutations.  This gives some ability to remove a specific mutation, but still preserves randomness in all circumstances.

Lorewise, this is a mixture of purifier and blob -- a semi-intelligent stem-cells treatment that will respond to the user's intentions.  No plans to make it manufacturable.

Adds a high profile finale room defending by 4 turrets and 1-2 secubots.  This is may an overkill level of difficulty, but the finale contains a single syringe with an equal chance of: alpha, medical, chimera, raptor, and smart purifier.

### CODE REVIEW

**Comment by Mecares** [comment]
I do not think that it is overkill, having a control laptop would trivialize it. I appreciate the recent changes to make labs more diversified. I would like a machine like a jammer to spawn on some lab floors that makes control laptops and the like unusable on that specific floor until deactivated or destroyed. That would make labs more challenging for me.

**Comment by BrianLefler** [comment]
No mutations -> Use item, "You feel cleansed" is the same effect purifier serum has.  Alternatively I could do a message like: "You have no need to use this." and keep the item.

Opinions on balance welcome:  1-2 untargeted mutations felt like the right price, but it could easily be 1 always, or 1-3.  Right now this is very rare, but you never know what someone'll add so I'd rather get the balance right at the outset.  My hope is that it's not a no-brainer to purify away all undesirable mutations, and you really should risk losing something good in the balance, but still be valuable enough that you'd be excited to find one of these.

**Comment by Night-Pryanik** [comment]
> Alternatively I could do a message like: "You have no need to use this." and keep the item.

That would be more logical imo.

**Comment by John-Candlebury** [comment]
The overkill reward seems rather dissapointing, the random nature of  mutagen and useless mutations makes the whole thing a coinflip. A single syringe could just as well just give you something terrible like nausea and fast metabolism and nothing good to tell for it.

Maybe if the mutagen were a variant that made you instantly cross the threshold?

**Comment by kevingranade** [comment]
Not gonna lie, I had a knee-jerk "nopenopenope" response to the targeted purifier (since it's been requested so many times), but you found a way to make it balanced that I didn't see, nice.

Not unique to this purifier version, but we should consider reevaluating "purifiable" status of mutations between mutation removals, i.e. right now, if you have mutations with dependencies like this:
```
A -> B -> C
       \> D
```
Only C and D are purifiable, but if C and D get removed during the process, then B should become a candidate.

**Comment by secretfire42** [comment]
My only objection to this - as someone whom has set up literally thousands of research labs - it would last less then a day past the setup. Academic researchers are lazy (and are more theoretical then practical) - they wouldn't bother to do this. Corporate ones might, until the security procedures cost them money. Military researchers are too practical. 

What I am saying is, one of the problems this brings up is the 'meta' perspective. If the game is going to take the aesthetic of simulating reality in a good way, the actual layout and scope should be realistic. You are not going to have a chemical that would see frequent use be this hard to access in the fashion you are describing. The actual balance aspect if it is pretty great, but the setup part is...eh. The labs themselves still feel pretty 'fake' as is for that reason; they are vastly improved and starting to feel more real, but in a gamey way; I think this is OK in a gamey game, but as the game moves more over towards simulation the aesthetic of it is increasingly galling.

So, instead of one obvious hazard, restricted (but presumably used for experiments on a regular basis) substance be guarded by....frankly kinda absurd security measures that no-one would implement IRL; what if you implement airlocks and such?

Take a look at this: https://www.popularmechanics.com/science/health/a6369/4315093/

In particular:

"To enter the restricted BSL-4 lab, Paragas first has to pass through two stainless-steel doors set up as an air lock. He punches in a code that deactivates the magnetic lock on the first door. The keypad also alerts the building automation system (BAS) that the air pressure is about to change. The BAS adjusts the airflow, increasing the pressure in the BSL-3 area and decreasing it in the air lock. Once Paragas is ready to enter BSL-4, the BAS will ensure that high-pressure air in the air lock flows into the low-pressure, high-security lab, trapping airborne pathogens. The deeper the level of containment, the lower the pressure. "Biocontainment sucks," Paragas says. "Any breach is sucked in by the negativity."

This is more in line with what I'd expect. It shouldn't be difficult to get to because ...there's a !@#!@ turret in there, it should be difficult to get to because you /cannot physically open the airlock/ unless certain requirements are met. This would be more annoying (I presume) from a coding factor to deal with but, I would suggest things like:

1. Power has to be restored to the airlock by hacking a certain terminal first. (which can be guarded)
2. The airlock itself will then have its own lock, with a code that must be recovered. This can be randomly placed on some piece of paper and distributed, either to that lab, or the nearest military base - but you'd have to search for it.
3. You'd also need to take off all your armor and be very encumbered, because the airlock is 'high pressure' and the guarded area would be 'low pressure' - which is to say, you couldn't breathe, and even if you could, it'd probably be infected by potential contaminants, as they wouldn't store just that one !@#@!# thing there. You might get ebola or something.
4. Heavily encumbered, with no armor or weapons, those scientist zombies and manhacks are going to be a much bigger threat now, huh? And what happens if they pierce your suit? Bad Things.
  
I'm probably overthinking it, but thats just my two cents.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/surgical-removal-of-extra-mutated-parts/17648/5

File: src/iuse.cpp [hunk 1]
```diff
@@ -1102,6 +1102,58 @@ int iuse::purify_iv( player *p, item *it, bool, const tripoint & )
...
+{
+    mutagen_attempt checks = mutagen_common_checks( *p, *it, false,
+                             pgettext( "memorial_male", "Injected smart purifier." ), pgettext( "memorial_female",
+                                     "Injected smart purifier." ) );
+    if( !checks.allowed ) {
+        return checks.charges_used;
+    }
+
+    std::vector<trait_id> valid; // Which flags the player has
+    std::vector<std::string> valid_names; // Which flags the player has
```
Comment (BevapDin) [inline src/iuse.cpp]: This is "which flags the player has" (according to the comment), but so is `valid` above. What's their difference, why are there two variables that - according to the comment - do / are the same?

**Pull request merged**